benchmark/perfdiff.cpp

```diff
@@ -95,20 +82,79 @@ int main(int argc, const char *argv[]) {
         ref.push_back(referenceThroughput);
     }
     // we check if the maximum of newcode is lower than minimum of ref, if so we have a problem so fail!
-    double worseref = *std::min_element(ref.begin(), ref.end());
-    double bestnewcode =  *std::max_element(newcode.begin(), newcode.end());
-    double bestref = *std::max_element(ref.begin(), ref.end());
-    double worsenewcode =  *std::min_element(newcode.begin(), newcode.end());
+    worseref = *std::min_element(ref.begin(), ref.end());
+    bestnewcode =  *std::max_element(newcode.begin(), newcode.end());
+    bestref = *std::max_element(ref.begin(), ref.end());
+    worsenewcode =  *std::min_element(newcode.begin(), newcode.end());
     std::cout << "The new code has a throughput in       " << worsenewcode << " -- " << bestnewcode << std::endl;
     std::cout << "The reference code has a throughput in " << worseref << " -- " << bestref << std::endl;
-    if(bestnewcode < worseref) {
-      std::cerr << "You probably have a performance degradation." << std::endl;
-      return EXIT_FAILURE;
+}
+
+
+int main(int argc, const char *argv[]) {
+    if (argc < 3) {
+        std::cerr << "Usage: " << argv[0] << " <old parse exe> <new parse exe> [<parse arguments>]" << std::endl;
+        return 1;
+    }
+
+    std::string newCommand = argv[1];
+    std::string refCommand = argv[2];
+    for (int i=3; i<argc; i++) {
+        newCommand += " ";
+        newCommand += argv[i];
+        refCommand += " ";
+        refCommand += argv[i];
     }
-    if(bestnewcode < worseref) {
-      std::cout << "You probably have a performance gain." << std::endl;
-      return EXIT_SUCCESS;
+    double worseref, bestref, worsenewcode, bestnewcode;
+    /**
+     * We take performance degradation seriously. When it occurs, we want
+     * to investigate it thoroughly. Theoretically, if INTERLEAVED_ATTEMPTS
+     * samples from one distribution are distinct from INTERLEAVED_ATTEMPTS
+     * from another distribution, then there should be a real difference.
+     * Unfortunately, in practice, we can get the impression that there are
+     * false positives. So the tool should make absolutely sure that the
+     * difference is entirely reproducible. So we require that it be
+     * able to reproduce it consistently MAX_TRIAL_COUNT times. Then it
+     * will be hard to argue with.
+     */
```

Comment:
I feel like we're going through more and more gyrations here, but I'm not a statistician :)

The idea here is that a group of 7 interleaved attempts is likely to give us a true best run from each, yes? I feel like it's a waste to lose that data as we move to the next attempt. But as long as this makes things better, I'm on board. It's a lot of false positives.

We do need to figure out how to make it quicker, though :/

Reply:
> I feel like we're going through more and more gyrations here, but I'm not a statistician 

Yeah. It is hackish.